GafferUI : Move render() from Gadget to ViewportGadget

danieldresser-ie opened the pull request:
This is an exploration of an API change that would make it easier for ViewportGadget to cache some extra data that could accelerate calls to render(). This is easier if we don't need to support render() from any Gadget, and it seems to make sense to just support it on ViewportGadget.

There are two potentially troublesome parts here:
* needing ViewportGadget to be a friend of Gadget ( since there are internal bits like m_glName that are needed during rendering )
* filterTypeId is not a good mechanism for selecting a subset of nodes to render ( particularly since I'm currently hardcoding to StandardConnectionGadget, whereas I guess I should be supporting any class derived from ConnectionGadget? ).   Any better ideas?

johnhaddon reviewed (commented):
This feels like a good direction to be going in. I wonder if we can go a bit further even?

- Can we remove `Gadget::renderRequestSignal()` and replace it with `ViewportGadget::renderRequestSignal()`? Since we can't render individual gadgets any more, it doesn't seem to make sense to have a signal per gadget?
- Can we remove `Gadget::select()`? If we can't render individual gadgets, and selection is coordinated by ViewportGadget, then it seems we should be able to hide this too?

> needing ViewportGadget to be a friend of Gadget

I think this is pretty reasonable. Gadget and ViewportGadget are parts of a single system, and friendship allows us to implement that system without exposing internals like `doRenderLayer()`, `select()` and `m_glName` publicly. In that sense it _increases_ encapsulation - it has enabled us to remove public APIs that didn't make much sense, and gains us more freedom in how we implement the internals.

> filterTypeId is not a good mechanism for selecting a subset of nodes to render

Agreed. I wonder if we even want it? It's a bit weird that `reconnectionGadgetAt()` will pick up connections that are completely obscured by nodes during normal rendering. Maybe we should just render everything? I suppose that'll be slower, but then with your forthcoming optimisations maybe that's not something we need to worry about?

danieldresser-ie commented on include/GafferUI/ViewportGadget.h:
```diff
@@ -168,6 +168,10 @@ class GAFFERUI_API ViewportGadget : public Gadget
 
+		/// A more flexible form of the above, this allows specifying a region to test instead of a point,
+		/// and optionally accepts filterTypeId - if set, only Gadgets with this type will be rendered	
+		void gadgetsAt( const Imath::Box2f &rasterRegion, std::vector<GadgetPtr> &gadgets, int filterTypeId = -1 ) const;
```
What should I do about these signatures?  It seems weird to have them not match?  Should I switch the existing gadgetsAt as well?

danieldresser-ie replied:
Oh, and I have updated to be using the enum values, though I'm now using the Layer enum rather than the typeId enum

johnhaddon replied:
> What should I do about these signatures? It seems weird to have them not match?

How about deprecating the old one, and adding an overload that matches the style of the new one?

danieldresser-ie commented on include/GafferUI/ViewportGadget.h:
```diff
@@ -237,6 +242,12 @@ class GAFFERUI_API ViewportGadget : public Gadget
 
+		void renderInternal( int filterTypeId = -1 ) const;
+
+		// Sets the GL state up with the name attribute and transform for
+		// this Gadget, makes sure the style is bound and then calls doRenderLayer().
```
Fixed in place.

danieldresser-ie commented on include/GafferUI/ViewportGadget.h:
```diff
@@ -237,6 +242,12 @@ class GAFFERUI_API ViewportGadget : public Gadget
+
+		// Sets the GL state up with the name attribute and transform for
+		// this Gadget, makes sure the style is bound and then calls doRenderLayer().
+		void renderLayer( const Gadget *gadget, Layer layer, const Style *currentStyle = nullptr, int filterTypeId = -1 ) const;
```
Fixed in place.

danieldresser-ie commented:
OK, I've had a go at addressing all feedback, other than the comment about the signature on gadgetsAt - should I switch the existing gadgetsAt signature while we're making API changes?

As for being able to filter what's rendered - we discussed just not filtering at all, which would have the desirable property of not being able to drop onto hidden noodles.  The problem is that the node we are dragging would itself block the selection.  The most correct would probably be to render everything but the current node - but I decided for the moment to go with something that is quick and keeps current behaviour, and kept the filter parameter, but made it based on the render Layer rather than Gadget id.  This makes a lot more sense than using Gadget id to me, and it should actually make things a bit faster.

Moving renderRequestSignal from Gadget to ViewportGadget totally makes sense, though it did end up being a few more changes than I expected.  I ended up adding a requestViewportRender method to Gadget, since finding the appropriate ViewportGadget and requesting a render is now a common thing.  Is that reasonable?

As for select(), I've kept the function in Gadget, since it seems like it should be paired with the Gadget constructor which sets up m_glName, but I've made it private.  Make sense?

johnhaddon reviewed (changes requested):
> The problem is that the node we are dragging would itself block the selection.

Doh! I should have thought of that!

> The most correct would probably be to render everything but the current node - but I decided for the moment to go with something that is quick and keeps current behaviour, and kept the filter parameter, but made it based on the render Layer rather than Gadget id.

Sounds good.

> I ended up adding a requestViewportRender method to Gadget, since finding the appropriate ViewportGadget and requesting a render is now a common thing. Is that reasonable?

Unless I've missed something, I don't think we want this. `dirty()` should be the one true way for a Gadget subclass to say that something has changed, and everything else should be hidden.

> As for select(), I've kept the function in Gadget, since it seems like it should be paired with the Gadget constructor which sets up m_glName, but I've made it private. Make sense?

Yep, makes sense. Although actually I think maybe we can do much better when we get to your `RenderItems` version. Can we pass the index within `m_renderItems` to `Selector::loadName()`, and then get to the gadget directly with `m_renderItems[hitRecord.name].gadget`? Or perhaps we're somehow relying on `NameStateComponent::glNameFromName()` to disambiguate against non-gadget names? It would be great to rid ourselves of that string munging if at all possible.

johnhaddon commented:
One additional note : we need to update Changes.md.

danieldresser-ie commented on src/GafferSceneUI/SceneGadget.cpp:
```diff
@@ -476,6 +476,12 @@ void SceneGadget::doRenderLayer( Layer layer, const GafferUI::Style *style ) con
 
+void SceneGadget::dirtyAll()
+{
+	// We only create this wrapper so it can be connected to updateRequiredSignal
```
Yep, much better.

danieldresser-ie commented on src/GafferUI/Gadget.cpp:
```diff
@@ -189,7 +189,7 @@ void Gadget::setVisible( bool visible )
 		Signals::emitLazily( m_signals.get(), &Signals::visibilityChangedSignal, this );
 	}
-	Signals::emitLazily( m_signals.get(), &Signals::renderRequestSignal, this );
+	requestViewportRender();
```
Yep, you're right.

danieldresser-ie commented on include/GafferUI/Gadget.h:
```diff
@@ -302,8 +300,9 @@ class GAFFERUI_API Gadget : public Gaffer::GraphComponent
 
-		/// \deprecated
-		void requestRender();
+		/// Triggers a render of the ViewportGadget containing this Gadget
+		void requestViewportRender();
```
Removed.

danieldresser-ie commented on src/GafferUIModule/GadgetBinding.cpp:
```diff
@@ -235,7 +233,7 @@ void GafferUIModule::bindGadget()
 	;
 
-	SignalClass<Gadget::RenderRequestSignal, DefaultSignalCaller<Gadget::RenderRequestSignal>, RenderRequestSlotCaller>( "RenderRequestSignal" );
+	SignalClass<Gadget::VisibilityChangedSignal, DefaultSignalCaller<Gadget::VisibilityChangedSignal>, VisibilityChangedSlotCaller>( "ButtonSignal" );
```
Fixed.

danieldresser-ie commented on include/GafferUI/Gadget.h:
```diff
@@ -307,6 +302,11 @@ class GAFFERUI_API Gadget : public Gaffer::GraphComponent
+		/// Returns the Gadget with the specified name, where name has been retrieved
+		/// from an IECoreGL::HitRecord after rendering some Gadget in GL_SELECT mode.
+		/// \todo Consider better mechanisms.
+		static GadgetPtr select( GLuint id );
```
I haven't bothered with this one, since it's private now, and I'm pretty sure you're idea to get rid of it my next PR will work.

danieldresser-ie commented:
I think I've addressed all feedback ( trickiest bit is "Simplify signature of gadgetsAt", since there were some knock-on effects of avoiding smart pointers. )

I've got the performance PR rebased onto this ( haven't uploaded that yet, since there's still lots of sawdust ), and it all seems to fit together pretty nicely now.  One annoying bit is that ViewportGadget needs to know when m_layoutDirty is set to false, and currently, in order to do this, I've added an extra method ViewportGadet::layoutDirty(), which is called explicitly from Gadget::dirty like renderRequestSignal is here ... not the prettiest, but I guess not really a problem if we've already decided there isn't much encapsulation between ViewportGadget and Gadget.

johnhaddon commented:
LGTM, except the `gadgetsAt()` bindings which were trying to return a raw pointer to Python. I've fixed that and tweaked Changes.md and will merge once CI goes through.
